### PR TITLE
Fix head change

### PR DESCRIPTION
**Pull request by elmattic**
## Summary of changes

Changes introduced in this pull request:

- Fix sticking mpool messages in case of "single tipset" errors
- Add unit test

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/state_manager/mod.rs [hunk 1]
```diff
@@ -1014,7 +1014,8 @@ where
                                 candidate_receipt = None;
                             }
                         }
-                        HeadChange::Apply(tipset) => {
+                        HeadChange::Apply(tipsets) => {
+                            let (tipset, _) = tipsets;
```
Comment (aatifsyed) [inline src/state_manager/mod.rs:1018]:
```suggestion
                        HeadChange::Apply((tipset, _)) => {
```

File: src/chain/store/chain_store.rs [hunk 2]
```diff
@@ -67,7 +67,7 @@ const DEFAULT_TIPSET_CACHE_SIZE: NonZeroUsize = nonzero!(8192usize);
 #[derive(Clone, Debug)]
 pub enum HeadChange {
     Current(Arc<Tipset>),
-    Apply(Arc<Tipset>),
+    Apply((Arc<Tipset>, i64)),
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply src/chain/store/chain_store.rs:70]: While we're on it - any reason this is a tuple as opposed to properties? Perhaps could be more self-explanatory like this: `Apple { to_apply: Arc<Tipset>, height: i64 }`, assuming I'm right about what those properties mean.

**Comment by lemmih** [comment]
Could you describe what the `head_change` function does? When the HEAD moves forward, it removes the messages that are included in the blockchain? And when the head moves backward, it adds the messages from those tipsets back into the message pool?

**Comment by lemmih** [comment]
Closing as there's no clear way forward.

**Pull request closed without merging**